### PR TITLE
[SPARK-30041][SQL][WEBUI] Add Codegen Stage Id to Stage DAG visualization in Web UI

### PR DESCRIPTION
**Pull request by LucaCanali**
### What changes were proposed in this pull request?
SPARK-29894 provides information on the Codegen Stage Id in WEBUI for SQL Plan graphs. Similarly, this proposes to add Codegen Stage Id in the DAG visualization for Stage execution. DAGs for Stage execution are available in the WEBUI under the Jobs and Stages tabs.

### Why are the changes needed?
This is proposed as an aid for drill-down analysis of complex SQL statement execution, as it is not always easy to match parts of the SQL Plan graph with the corresponding Stage DAG execution graph. Adding Codegen Stage Id for WholeStageCodegen operations makes this task easier.

### Does this PR introduce any user-facing change?
Stage DAG visualization in the WEBUI will show codegen stage id for WholeStageCodegen operations, as in the example snippet from the WEBUI, Jobs tab  (the query used in the example is TPCDS 2.4 q14a):

### How was this patch tested?
Manually tested, see also example snippet.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala [hunk 1]
```diff
@@ -205,7 +205,14 @@ abstract class SparkPlan extends QueryPlan[SparkPlan] with Logging with Serializ
    * for visualization.
    */
   protected final def executeQuery[T](query: => T): T = {
-    RDDOperationScope.withScope(sparkContext, nodeName, false, true) {
+    val nodeNameScope =
+      if (nodeName == "WholeStageCodegen") {
```
Comment (konjac) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala]: It is not a good idea to use string comparison. Why not using type match similar to [PR-26519](https://github.com/apache/spark/pull/26519/files)?

Comment (konjac) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala]: How about making nodeNameScope a property in TreeNode?

Comment (LucaCanali) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala]: Indeed match expression is a better way to do this, thanks.
I am not sure about modifying TreeNode, I see this proposed implementation as having a smaller footprint.

Comment (konjac) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala]:
```scala
    val nodeNameScope = this match {
      case plan: WholeStageCodegenExec => s"${plan.nodeName} (${plan.codegenStageId})"
      case _ => nodeName
    }
```

**Comment by LucaCanali** [comment]
@cloud-fan, this follows from #26519 and it is also aimed at making it easier to understand SQL execution and plans via the WEB UI. You may be interested to have a look.

**Comment by cloud-fan** [comment]
Can we have some screenshots?

**Comment by cloud-fan** [comment]
ok to test

**Comment by cloud-fan** [comment]
cc @gengliangwang

**Comment by LucaCanali** [comment]
Any additional comments on this? From the logs I'd say Jenkins failure of the latest test was transient?

**Comment by tgravescs** [comment]
test this please

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
retest this please

**Comment by cloud-fan** [comment]
@HeartSaVioR `org.apache.spark.sql.kafka010.KafkaDelegationTokenSuite` seems still flaky. Shall we create a JIRA ticket to investigate it further?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala [hunk 2]
```diff
@@ -206,7 +206,11 @@ abstract class SparkPlan extends QueryPlan[SparkPlan] with Logging with Serializ
    * for visualization.
    */
   protected final def executeQuery[T](query: => T): T = {
-    RDDOperationScope.withScope(sparkContext, nodeName, false, true) {
+    val nodeNameScope = this match {
```
Comment (gengliangwang) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala]: How about overriding the `nodeName` of `WholeStageCodegenExec`.
I think it is hacky to make special handling here.

**Comment by HeartSaVioR** [comment]
@cloud-fan 

> @HeartSaVioR org.apache.spark.sql.kafka010.KafkaDelegationTokenSuite seems still flaky. Shall we create a JIRA ticket to investigate it further?

Thanks for pinging. I'll create a new JIRA issue to track this.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala (hunk 2 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala]: ah good idea! we can also remove the hack in SPARK-29894

Comment (LucaCanali) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala]: +1

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**